cypress-io/cypress
# chore: add GraphQL mutation for sending system notifications via Electron

- Closes https://github.com/cypress-io/cypress/issues/26687

### Additional details


This PR adds a new GraphQL mutation and Electron action that allows us to send a system notification to the OS using the [Electron notifications API](https://www.electronjs.org/docs/latest/tutorial/notifications). We've manually tested these notifications on macOS, Linux, and Windows operating systems:

**macOS**

https://www.loom.com/share/18a187afab884957af3245508a0f3de3

**Windows**

**OUTSTANDING ISSUE**
There is an issue on Windows where when a user clicks the action to turn off notifications for Cypress, the Cypress app does not appear in the Windows notification settings. This means that users cannot re-enable notifications for the Cypress app once they have been disabled this way at the OS level.

I suspect that this has to do with the way that Cypress is installed. Most applications have a Windows installer that sets up shortcuts for the applications, but Cypress doesn't do this because it is launched from the command line. We are discussing how to work around this or find a solution.

Created https://github.com/cypress-io/cypress/issues/26786 to deal with this

https://www.loom.com/share/faf661068e6d4391b128aa4525cd3054

**Ubuntu**

https://github.com/cypress-io/cypress/assets/14275198/c038efc6-2918-4bcc-8f55-4ff51c2f329a

### Steps to test


You can test the notifications using the GraphiQL GUI

1. Run Cypress
2. Open a project in a browser
3. With Launchpad focused, click "Developer Tools" in the menu bar and select "GraphiQL"
4. Write a mutation
```
mutation showNotification($title: String!, $body: String) {
  showSystemNotification(title: $title, body: $body)
}
```
5. Add variables
```json
{
  "title": "My notification",
  "body": "This is a test notification"
}
```
6. Press the play button to execute the query
7. Verify that Cypress sent a notification to your operating system (take action to allow notifications if needed)
8. Verify that clicking on the notification focuses your active browser window (not Launchpad)

### How has the user experience changed?


This should not change the user experience because this mutation is not being executed anywhere in the app.

### PR Tasks


- [na] Have tests been added/updated? **I decided that it wouldn't be worth writing automated tests for this yet since it's not called yet. Once we use it somewhere, we can test it at that level instead**
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
